---
date created: 2022-01-23 15:09
date modified: 2022-01-23 15:09
title: 测试
---
rust 测试分为两种：
- unit tests：单元测试通常在同一个模块下，使用 `#[cfg(test)]` 属性，将所有与测试相关的代码放在此属性下面。
- integration tests：交互测试，建立一个与 src 目录同级的文件夹 **tests**

## 测试相关的宏
> [!note]
> - `#[name]`：针对测试函数的注解属性
> - `#![name]`：会应用在整个 crate，通常被使用在crate root 顶部。
> - `#[cfg(test)]`：此属性会添加一个顶层的test模块，并让编译器只在 test 模式下才会编译。
> - `assert!`
>> [!example]
>> ```rust
>> assert!(a==b,"{} was not equal to {}",a,b);
>> assert_eq!(a,b,"{} and {} are not equal",a,b);
>> assert_ne!
>> // debug的宏，仅在 debug 模式
>> - `debug_assert!`
>> - `debug_assert_eq!`
>> - `debug_assert_ne!`

为了使自定义的类型能使用上面的宏，需要在类型上使用 `#[derive（Eq,PartialEq)]` 注解。

> [!note] 隔离测试代码，使用注解 `#[cfg(test)]`
> ```rust
> fn sum(a: i8, b: i8) -> i8 {
 >   a + b
> }
> #[cfg(test)]
> mod tests {
 >   fn sum_inputs_outputs() -> Vec<((i8, i8), i8)> {
>        vec![((1, 1), 2), ((0, 0), 0), ((2, -2), 0)]
>    }
>  #[test]
>    fn test_sums() {
 >       for (input, output) in sum_inputs_outputs() {
>            assert_eq!(crate::sum(input.0, input.1), output);
>        }
>    }
> }


> [!note] 失败测试
> `#[should_panic]` 表示函数会引起一个不会恢复的错误
> ```rust
> #[test]
> #[should_panic]
> fn this_panic() {
>	assert_eq!(1,2);
> }


> [!note] 忽略测试 `#[ignore]`
> ```rust
> #[cfg(test)]
> mod tests {
>	#[test]
>	#[ignore]
>	fn ignore_func() {
>		assert_eq!(true);
>	}
>}


## benchmark
`#[bench]`：注解表示该函数是一个 benchmark test

> [!example]
> ```rust
> #![feature(test)]
> exter carte test;
>
> use test::Bencher;
>
> pub fn do_nothing_slowly() {
>	print!(".");
>	for _ in 1..10_000_000 {};
> }
>
> pub fn do_nothing_fast() {}
>
> #[bench]
> fn bench_nothing_slowly(b: &mut Bencher) {
> 	b.iter(|| do_nothing_slowly());
> }
>
> #[bench]
> fn bench_nothing_fast(b: &mut Bencher) {
>	b.iter(|| do_nothing_fast());
> }
